README.md
Wordpress Editorial-Control Plugin
===========================

Upgrades the contributor role in WordPress so that any contributor can also upload images and pictures. A contributor is now made into a super-contributor.

### Features
* Email Notifications for Each Stage (Approval Emails in TXT & HTML)
* WP-Updates Support
* Presstrends Support
* Admin Options Page
* Config in Plugin to setup defaults
* Embed HTML on/off for Contributors

### Copyright & Licence
Copyright 2013 Vincent Koc (https://github.com/koconder)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

CHANGELOG.md
#### Changelog
##### Foreword
Below is a rougth changelog to document changes to the plugin

##### Changes
v3.2.1 date: 2014-04-09

		- Show version number in Settings panel in footer

v3.2.0 date: 2014-04-09

		- HTML Filtering and Embed Capabillity given to Role
		- Split out the plugin code for extenabillity
		- Security to disable direct access to folder or files
		- Created a Changelog, Reademe Markdown files
		- Renamed functions to match naming conventions
		- Updated copyright information
		- Presstrends Enable/Disable in Config
		- Default editor email handeling

v3.1.0 date: 2014-03-06

		- Inital Commit
		- Added Admin Section
		- Added Wp-Updates Auto Updater
		- Added Presstrends
		- Added Email TXT & HTML Formats


inc-options.php
<?php
// If this file is called directly, abort.
if (!defined('WPINC')){
    die;
}
?>
<div class="wrap"><h2>Editorial Control Status</h2>
     <form name="site" action="" method="post" id="notifier">
         <div id="review">
             <fieldset id="pendingdiv">
                 <legend><b><?php _e('Editorial Control settings are set to:') ?></b></legend>
             </fieldset>
             <br />
             <fieldset id="reviewdiv">
                 <div>
                     <input type="text" size="50" name="notificationemails" tabindex="1" id="notificationemails" value="<?php echo attribute_escape(get_option('notificationemails')); ?>"><br />
                     Enter the email address of who should be notified when a post is in a pending review status(comma separated).
                 </div>
                 <br/>
                 <div>
                     <label for="supercontributor" class="selectit">
                         <input type="checkbox" tabindex="2" id="supercontributor" name="supercontributor" value="yes" <?php if(get_option('supercontributor')=='yes') echo 'checked="checked"'; ?> />
                         Allow the contributor role to upload images.
                     </label>
                     <br/>
                     <label for="contributorembed" class="selectit">
                         <input type="checkbox" tabindex="2" id="contributorembed" name="contributorembed" value="yes" <?php if(get_option('contributorembed')=='yes') echo 'checked="checked"'; ?> />
                         Allow the contributor role to place javascript and embeds.
                     </label>
                     <br/>
                     <label for="approvednotification" class="selectit">
                         <input type="checkbox" tabindex="2" id="approvednotification" name="approvednotification" value="yes" <?php if(get_option('approvednotification')=='yes') echo 'checked="checked"'; ?> />
                         Notify the contributor when their post is approved.
                     </label>
                     <br />
                     <label for="declinednotification" class="selectit">
                         <input type="checkbox" tabindex="3" id="declinednotification" name="declinednotification" value="yes" <?php if(get_option('declinednotification')=='yes') echo 'checked="checked"'; ?> />
                         Notify contributor when their post is declined (set back to draft).
                     </label>
                 </div>
             </fieldset>
             <br />
             <p class="submit">
                 <input name="save" type="submit" id="savenotifier" tabindex="6" style="font-weight: bold;" value="Save Settings" />
             </p>
         </div>
     </form>
     <small>Powered by WP Editorial Control</small>
     <small>v<?php echo plugin_get_version() ?></small>
</div>

include/class.php
<?php

/**
 * Plugin - Default Settings
 */
function editorial_control_defaults(){
	add_option('supercontributor', 'yes');
	update_option('supercontributor', 'yes');
	allow_contributor_uploads();

	// add_action('admin_init', 'allow_contributor_uploads');
	// if ( current_user_can('contributor') && !current_user_can('upload_files') ) {
	// }

	add_option('notificationemails', 'editor@pokerstars.com');
	add_option('approvednotification', 'yes');
	add_option('declinednotification', 'yes');
	$from_email = parse_url(get_option('siteurl') , PHP_URL_HOST);
	$from_email = ltrim($from_email, 'www.');
	$from_email = "admin@" . $from_email;
	add_option('fromemail', $from_email);
}

/**
 * Plugin - Deactivate
 */
function editorial_control_deactivate(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
	update_option('supercontributor', '');
}

/**
 * Plugin - Remove
 */
function editorial_control_uninstall(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
	delete_option('supercontributor');
	delete_option('notificationemails');
	delete_option('approvednotification');
	delete_option('declinednotification');
	delete_option('fromemail');
}

/**
 * Menu Hook - Add Sub-Menu
 */
function ec_add_option_page()
{
	add_options_page('Editorial Control', 'Editorial Control', 'edit_themes', 'editorial_control', 'ec_options_page');
}

/**
 * Contributor Uploads - Allow
 */
function allow_contributor_uploads(){
	$contributor = get_role('contributor');
	$contributor->add_cap('upload_files');
}

/**
 * Contributor Uploads - Disallow
 */
function disallow_contributor_uploads(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
}

/**
 * Contributor Uploads - Disallow
 */
function presstrends_plugin(){
	//Load Config
	global $ec_config;
	
	// PressTrends Account API Key
	$api_key = $ec_config['presstrends_api_key'];
	$auth = $ec_config['presstrends_api_auth'];

	// Start of Metrics
	global $wpdb;
	
	$data = get_transient('presstrends_cache_data');
	if (!$data || $data == '') {
		$api_base = 'http://api.presstrends.io/index.php/api/pluginsites/update?auth=';
		$url = $api_base . $auth . '&api=' . $api_key . '';
		$count_posts = wp_count_posts();
		$count_pages = wp_count_posts('page');
		$comments_count = wp_count_comments();
		if (function_exists('wp_get_theme')) {
			$theme_data = wp_get_theme();
			$theme_name = urlencode($theme_data->Name);
		}
		else {
			$theme_data = get_theme_data(get_stylesheet_directory() . '/style.css');
			$theme_name = $theme_data['Name'];
		}

		$plugin_name = '&';
		foreach(get_plugins() as $plugin_info) {
			$plugin_name.= $plugin_info['Name'] . '&';
		}

		// CHANGE __FILE__ PATH IF LOCATED OUTSIDE MAIN PLUGIN FILE
		$plugin_data = get_plugin_data(__FILE__);
		$posts_with_comments = $wpdb->get_var("SELECT COUNT(*) FROM $wpdb->posts WHERE post_type='post' AND comment_count > 0");
		$data = array(
			'url' => base64_encode(site_url()) ,
			'posts' => $count_posts->publish,
			'pages' => $count_pages->publish,
			'comments' => $comments_count->total_comments,
			'approved' => $comments_count->approved,
			'spam' => $comments_count->spam,
			'pingbacks' => $wpdb->get_var("SELECT COUNT(comment_ID) FROM $wpdb->comments WHERE comment_type = 'pingback'") ,
			'post_conversion' => ($count_posts->publish > 0 && $posts_with_comments > 0) ? number_format(($posts_with_comments / $count_posts->publish) * 100, 0, '.', '') : 0,
			'theme_version' => $plugin_data['Version'],
			'theme_name' => $theme_name,
			'site_name' => str_replace(' ', '', get_bloginfo('name')) ,
			'plugins' => count(get_option('active_plugins')) ,
			'plugin' => urlencode($plugin_name) ,
			'wpversion' => get_bloginfo('version') ,
		);
		foreach($data as $k => $v) {
			$url.= '&' . $k . '=' . $v . '';
		}

		wp_remote_get($url);
		set_transient('presstrends_cache_data', $data, 60 * 60 * 24);
	}
}

/**
 * Admin Page Functions
 */
function ec_options_page(){
	//Check if we have a save
	if(isset($_POST['save'])){
		//Update the Super Contibutor
		update_option('supercontributor', $_POST['supercontributor']);
		
		//Check Allow-Disallow Uploads
		if (isset($_POST['supercontributor'])) {
			allow_contributor_uploads();
		}
		else {
			disallow_contributor_uploads();
		}

		//Update Notification Emails
		update_option('notificationemails', $_POST['notificationemails']);
		update_option('approvednotification', $_POST['approvednotification']);
		update_option('declinednotification', $_POST['declinednotification']);
		
		//Show Message
		echo "<div id='message' class='updated fade'><p>Editorial control settings have now been saved.</p></div>";
	}
	
	//Load Admin Page
	include('options.php');	
}


/**
 * Email Notifications Function
 */
function notify_status($new_status, $old_status, $post){
	global $current_user;
	
	$contributor = get_userdata($post->post_author);
	$boundary = uniqid();
	
	if ($old_status != 'pending' && $new_status == 'pending') {
		$emails = get_option('notificationemails');
		if (strlen($emails)) {
			$post_title = get_the_title($post->ID);
			$from_email = get_option('fromemail');
			$headers = 'From : Review Post <' . $from_email . '>' . "\r\n" . "MIME-Version: 1.0\r\n" . "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post_title . '" is pending review';
			$message = "This is a MIME encoded message.";
			$message.= "\r\n\r\n--" . $boundary . "\r\n";
			$message.= "Content-type: text/plain;charset=utf-8\r\n\r\n";
			$message.= "
                A new post on [" . get_option('blogname') . "] by {$contributor->display_name} is pending review. \n
                URL : " . get_option('siteurl') . ". \n
                Author : {$contributor->user_login} <{$contributor->user_email}> (IP:{$_SERVER['REMOTE_ADDR']}). \n
                Title : $post_title. \n
                Powered by: WP Editorial Control
            ";
			$message.= "\r\n\r\n--" . $boundary . "\r\n";
			$message.= "Content-type: text/html;charset=utf-8\r\n\r\n";
			$message.= "
                <strong>A new post on [" . get_option('blogname') . "] by {$contributor->display_name} is pending review.</strong>
                <br/><br/>
                URL : " . get_option('siteurl') . "
                <br/>
                Author : {$contributor->user_login} <{$contributor->user_email}> (IP:{$_SERVER['REMOTE_ADDR']})
                <br/>
                Title : $post_title
            ";
			$category = get_the_category($post->ID);
			if (isset($category[0])) {
				$message.= " (Category : {$category[0]->name})<br/><br/>";
			}

			$message.= "
                Preview it: " . get_option('siteurl') . "/?p={$post->ID}&preview=true
                <br/>
                Edit/Publish it: " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID}
                <br/>
            ";
			$message.= "<br/>###";
			$post = get_post($post->ID);
			$message.= apply_filters('the_content', $post->post_content);
			$message.= "###<br/><br/>";
			$message.= "Edit/Publish it: " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID}<br/><br/><br/>";
			$message.= "Powered by: WP Editorial Control";
			$message.= "\r\n\r\n--" . $boundary . "--";
			wp_mail($emails, $subject, $message, $headers);
		}
	}
	elseif ($old_status == 'pending' && $new_status == 'publish' && $current_user->ID != $contributor->ID) {
		if (get_option('approvednotification') == 'yes') {
			$from_email = get_option('fromemail');
			$headers[] = 'From: Review Post <' . $from_email . '>';
			$headers[] = "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post->post_title . '" approved';
			$message = "{$contributor->display_name},\n\nCongratulations, your post has been approved and published at " . get_permalink($post->ID) . " .\n\n";
			$message.= "By {$current_user->display_name} <{$current_user->user_email}>\n\n\n";
			$message.= "Powered by: WP Editorial Control";
			wp_mail($contributor->user_email, $subject, $message, $headers);
		}
	}
	elseif ($old_status == 'pending' && $new_status == 'draft' && $current_user->ID != $contributor->ID) {
		if (get_option('declinednotification') == 'yes') {
			$from_email = get_option('fromemail');
			$headers[] = 'From: Review Post <' . $from_email . '>';
			$headers[] = "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post->post_title . '" declined';
			$message = "{$contributor->display_name},\n\nSorry ,our post has not been approved. You can edit the post at " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID} .\n\n";
			$message.= "By {$current_user->display_name} <{$current_user->user_email}>\n\n\n";
			$message.= "Powered by: Editorial Control";
			wp_mail($contributor->user_email, $subject, $message, $headers);
		}
	}
}

editorial-control.php
<?php
/**
 * Editorial Control UK
 *
 * Upgrade the contributor role in WordPress so that any contributor can also upload images and pictures.
 * Sends email notification when posts are pending review, with fixes for email layouts.
 *
 * @package   Wordpress-Editorial-Control
 * @author    Onder Vincent Koc <vincent@loophole.eu>
 * @license   GPL-2.0+
 * @link      https://github.com/koconder/Wordpress-Editorial-Control
 * @copyright 2014 Onder Vincent Koc
 *
 * @wordpress-plugin
 * Plugin Name:       Editorial Control UK
 * Plugin URI:        https://github.com/koconder
 * Description:       Upgrade the contributor role in WordPress so that any contributor can also upload images and pictures and sends an email notification
 * Version:           3.2.0
 * Author:            Onder Vincent Koc
 * Author URI:        https://github.com/koconder
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * 
 * Copyright 2013 Vincent Koc (https://github.com/koconder)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */

// Global Variable Config ~ https://codex.wordpress.org/Function_Reference/register_activation_hook
global $ec_config;
 
// If this file is called directly, abort.
if (!defined('WPINC')){
    die;
}

/*----------------------------------------------------------------------------*
 * SETUP
 *----------------------------------------------------------------------------*/
$ec_config = array();
$ec_config['presstrends_api_key'] = false;//PressTrends API Key
$ec_config['presstrends_api_auth'] = false;//PressTrends AUTH Key
$ec_config['default_editor_email'] = false;//Editor Email

/*----------------------------------------------------------------------------*
 * Includes
 *----------------------------------------------------------------------------*/
require_once('inc-class.php');

/*----------------------------------------------------------------------------*
 * WP-Updates
 *----------------------------------------------------------------------------*/
require_once('wp-updates-plugin.php');
new WPUpdatesPluginUpdater_297( 'http://wp-updates.com/api/2/plugin', plugin_basename(__FILE__));

// This is for testing only!
//set_site_transient( 'update_plugins', null );

// Show which variables are being requested when query plugin API
//add_filter( 'plugins_api_result', array(&amp;$this, 'debug_result'), 10, 3 );

/*----------------------------------------------------------------------------*
 * Activation - Deactivation Hooks
 *----------------------------------------------------------------------------*/

// To set default config on activation
register_activation_hook(__FILE__,'editorial_control_defaults');

// To turn off supercontributor if de-activated
register_deactivation_hook( __FILE__, 'editorial_control_deactivate' );

// To remove option flag when uninstalled
register_uninstall_hook(__FILE__,'editorial_control_uninstall');


/*----------------------------------------------------------------------------*
 * Post - Email Hooks
 *----------------------------------------------------------------------------*/
// Hook for post status changes
add_filter('transition_post_status', 'editorial_control_notify_status',10,3);

/*----------------------------------------------------------------------------*
 * Admin and Analytics
 *----------------------------------------------------------------------------*/
// Hook for adding admin menus
add_action('admin_menu', 'editorial_control_add_option_page');

// PressTrends WordPress Action
if(!empty($ec_config['presstrends_api_key']) && !empty($ec_config['presstrends_api_auth'])){
	add_action('admin_init', 'presstrends_plugin');
}


wp-updates-plugin.php
<?php
/*
WPUpdates Plugin Updater Class
http://wp-updates.com
v2.0

Example Usage:
require_once('wp-updates-plugin.php');
new WPUpdatesPluginUpdater_297( 'http://wp-updates.com/api/2/plugin', plugin_basename(__FILE__) );
*/

if( !class_exists('WPUpdatesPluginUpdater_297') ) {
    class WPUpdatesPluginUpdater_297 {
    
    	var $api_url;
    	var $plugin_id = 297;
    	var $plugin_path;
    	var $plugin_slug;
    	var $license_key;
    
    	function __construct( $api_url, $plugin_path, $license_key = null ) {
    		$this->api_url = $api_url;
    		$this->plugin_path = $plugin_path;
    		$this->license_key = $license_key;
    		if(strstr($plugin_path, '/')) list ($t1, $t2) = explode('/', $plugin_path); 
    		else $t2 = $plugin_path;
    		$this->plugin_slug = str_replace('.php', '', $t2);
    
    		add_filter( 'pre_set_site_transient_update_plugins', array(&$this, 'check_for_update') );
    		add_filter( 'plugins_api', array(&$this, 'plugin_api_call'), 10, 3 );
    		
    		// This is for testing only!
    		//set_site_transient( 'update_plugins', null );
    
    		// Show which variables are being requested when query plugin API
    		//add_filter( 'plugins_api_result', array(&$this, 'debug_result'), 10, 3 );
    	}
    
    	function check_for_update( $transient ) {
    		if(empty($transient->checked)) return $transient;
    		
    		$request_args = array(
    		    'id' => $this->plugin_id,
    		    'slug' => $this->plugin_slug,
    			'version' => $transient->checked[$this->plugin_path]
    		);
    		if ($this->license_key) $request_args['license'] = $this->license_key;
    		
    		$request_string = $this->prepare_request( 'update_check', $request_args );
    		$raw_response = wp_remote_post( $this->api_url, $request_string );
    		
    		$response = null;
    		if( !is_wp_error($raw_response) && ($raw_response['response']['code'] == 200) )
    			$response = unserialize($raw_response['body']);
    		
    		if( is_object($response) && !empty($response) ) {
                // Feed the update data into WP updater
    			$transient->response[$this->plugin_path] = $response;
                return $transient;
            }

            // Check to make sure there is not a similarly named plugin in the wordpress.org repository
            if ( isset( $transient->response[$this->plugin_path] ) ) {
                if ( strpos( $transient->response[$this->plugin_path]->package, 'wordpress.org' ) !== false  ) {
                    unset($transient->response[$this->plugin_path]);
                }
            }

    		return $transient;
    	}
    
    	function plugin_api_call( $def, $action, $args ) {
    		if( !isset($args->slug) || $args->slug != $this->plugin_slug ) return $def;
    		
    		$plugin_info = get_site_transient('update_plugins');
    		$request_args = array(
    		    'id' => $this->plugin_id,
    		    'slug' => $this->plugin_slug,
    			'version' => (isset($plugin_info->checked)) ? $plugin_info->checked[$this->plugin_path] : 0 // Current version
    		);
    		
    		$request_string = $this->prepare_request( $action, $request_args );
    		$raw_response = wp_remote_post( $this->api_url, $request_string );
    		
    		if( is_wp_error($raw_response) ){
    			$res = new WP_Error('plugins_api_failed', __('An Unexpected HTTP Error occurred during the API request.</p> <p><a href="?" onclick="document.location.reload(); return false;">Try again</a>'), $raw_response->get_error_message());
    		} else {
    			$res = unserialize($raw_response['body']);
    			if ($res === false)
    				$res = new WP_Error('plugins_api_failed', __('An unknown error occurred'), $raw_response['body']);
    		}
    		
    		return $res;
    	}
    
    	function prepare_request( $action, $args ) {
    		global $wp_version;
    		
    		return array(
    			'body' => array(
    				'action' => $action, 
    				'request' => serialize($args),
    				'api-key' => md5(home_url())
    			),
    			'user-agent' => 'WordPress/'. $wp_version .'; '. home_url()
    		);	
    	}
    	
    	function debug_result( $res, $action, $args ) {
    		echo '<pre>'.print_r($res,true).'</pre>';
    		return $res;
    	}
    
    }
}

?>

inc-class.php
<?php
// If this file is called directly, abort.
if (!defined('WPINC')){
    die;
}

/**
 * Plugin - Default Settings
 */
function editorial_control_defaults(){
	//Load Config
	global $ec_config;
	
	//Load Roles Super Contibutor
	add_option('supercontributor', 'yes');
	update_option('supercontributor', 'yes');
	allow_contributor_uploads();
	
	//Load Roles HTML Embeds
	add_option('contributorembed', 'yes');
	update_option('contributorembed', 'yes');
	allow_contributor_embeds();

	//Build Site Email
	$site_domain = parse_url(get_option('siteurl') , PHP_URL_HOST);
	$site_domain = ltrim($site_domain, 'www.');

	//Email Settings
	if(empty($ec_config['default_editor_email'])){
		//If we have nothing set sending to editor@{wordpressdomain}.com
		$ec_config['default_editor_email'] = 'editor@' . $site_domain;
	}
	add_option('notificationemails', $ec_config['default_editor_email']);
	add_option('approvednotification', 'yes');
	add_option('declinednotification', 'yes');
	add_option('fromemail', 'admin@'.$site_domain);

	// Test
	// add_action('admin_init', 'allow_contributor_uploads');
	// if ( current_user_can('contributor') && !current_user_can('upload_files') ) {
	// }
}

/**
 * Plugin - Deactivate
 */
function editorial_control_deactivate(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
	$contributor->remove_cap('unfiltered_html');
	update_option('supercontributor', '');
	update_option('contributorembed', '');
}

/**
 * Plugin - Remove
 */
function editorial_control_uninstall(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
	$contributor->remove_cap('unfiltered_html');
	delete_option('supercontributor');
	delete_option('notificationemails');
	delete_option('approvednotification');
	delete_option('declinednotification');
	delete_option('fromemail');
}

/**
 * Menu Hook - Add Sub-Menu
 */
function editorial_control_add_option_page()
{
	add_options_page('Editorial Control', 'Editorial Control', 'edit_themes', 'editorial_control', 'editorial_control_options_page');
}

/**
 * Contributor Uploads - Allow
 */
function allow_contributor_uploads(){
	$contributor = get_role('contributor');
	$contributor->add_cap('upload_files');
}

/**
 * Contributor Uploads - Disallow
 */
function disallow_contributor_uploads(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('upload_files');
}

/**
 * HTML Embeds - Allow
 */
function allow_contributor_embeds(){
	$contributor = get_role('contributor');
	$contributor->add_cap('unfiltered_html');
}

/**
 * HTML Embeds - Disallow
 */
function disallow_contributor_embeds(){
	$contributor = get_role('contributor');
	$contributor->remove_cap('unfiltered_html');
}

/**
 * Press Trends
 */
function presstrends_plugin(){
	//Load Config
	global $ec_config;
	
	// PressTrends Account API Key
	$api_key = $ec_config['presstrends_api_key'];
	$auth = $ec_config['presstrends_api_auth'];

	// Start of Metrics
	global $wpdb;
	
	$data = get_transient('presstrends_cache_data');
	if (!$data || $data == '') {
		$api_base = 'http://api.presstrends.io/index.php/api/pluginsites/update?auth=';
		$url = $api_base . $auth . '&api=' . $api_key . '';
		$count_posts = wp_count_posts();
		$count_pages = wp_count_posts('page');
		$comments_count = wp_count_comments();
		if (function_exists('wp_get_theme')) {
			$theme_data = wp_get_theme();
			$theme_name = urlencode($theme_data->Name);
		}
		else {
			$theme_data = get_theme_data(get_stylesheet_directory() . '/style.css');
			$theme_name = $theme_data['Name'];
		}

		$plugin_name = '&';
		foreach(get_plugins() as $plugin_info) {
			$plugin_name.= $plugin_info['Name'] . '&';
		}

		// CHANGE __FILE__ PATH IF LOCATED OUTSIDE MAIN PLUGIN FILE
		$plugin_data = get_plugin_data(__FILE__);
		$posts_with_comments = $wpdb->get_var("SELECT COUNT(*) FROM $wpdb->posts WHERE post_type='post' AND comment_count > 0");
		$data = array(
			'url' => base64_encode(site_url()) ,
			'posts' => $count_posts->publish,
			'pages' => $count_pages->publish,
			'comments' => $comments_count->total_comments,
			'approved' => $comments_count->approved,
			'spam' => $comments_count->spam,
			'pingbacks' => $wpdb->get_var("SELECT COUNT(comment_ID) FROM $wpdb->comments WHERE comment_type = 'pingback'") ,
			'post_conversion' => ($count_posts->publish > 0 && $posts_with_comments > 0) ? number_format(($posts_with_comments / $count_posts->publish) * 100, 0, '.', '') : 0,
			'theme_version' => $plugin_data['Version'],
			'theme_name' => $theme_name,
			'site_name' => str_replace(' ', '', get_bloginfo('name')) ,
			'plugins' => count(get_option('active_plugins')) ,
			'plugin' => urlencode($plugin_name) ,
			'wpversion' => get_bloginfo('version') ,
		);
		foreach($data as $k => $v) {
			$url.= '&' . $k . '=' . $v . '';
		}

		wp_remote_get($url);
		set_transient('presstrends_cache_data', $data, 60 * 60 * 24);
	}
}

/**
 * Admin Page Functions
 */
function editorial_control_options_page(){
	//Check if we have a save
	if(isset($_POST['save'])){
		//Update the Super Contibutor
		update_option('supercontributor', $_POST['supercontributor']);
		if (isset($_POST['supercontributor'])) {
			allow_contributor_uploads();
		}
		else {
			disallow_contributor_uploads();
		}

		//Update the HTML Embed
		update_option('contributorembed', $_POST['contributorembed']);
		if (isset($_POST['contributorembed'])) {
			allow_contributor_embeds();
		}
		else {
			disallow_contributor_embeds();
		}

		//Update Notification Emails
		update_option('notificationemails', $_POST['notificationemails']);
		update_option('approvednotification', $_POST['approvednotification']);
		update_option('declinednotification', $_POST['declinednotification']);
		
		//Show Message
		echo "<div id='message' class='updated fade'><p>Editorial control settings have now been saved.</p></div>";
	}
	
	//Load Admin Page
	include('inc-options.php');
}

/**
 * Returns current plugin version for admin page.
 * 
 * @return string Plugin version
 * @link	http://code.garyjones.co.uk/get-wordpress-plugin-version
 */
function plugin_get_version(){
	if ( ! function_exists( 'get_plugins' ) )
		require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
	$plugin_folder = get_plugins( '/' . plugin_basename( dirname( __FILE__ ) ) );
	$plugin_file = basename( ( __FILE__ ) );
	return $plugin_folder[$plugin_file]['Version'];
}

/**
 * Email Notifications Function
 */
function editorial_control_notify_status($new_status, $old_status, $post){
	global $current_user;
	
	$contributor = get_userdata($post->post_author);
	$boundary = uniqid();
	
	if ($old_status != 'pending' && $new_status == 'pending') {
		$emails = get_option('notificationemails');
		if (strlen($emails)) {
			$post_title = get_the_title($post->ID);
			$from_email = get_option('fromemail');
			$headers = 'From : Review Post <' . $from_email . '>' . "\r\n" . "MIME-Version: 1.0\r\n" . "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post_title . '" is pending review';
			$message = "This is a MIME encoded message.";
			$message.= "\r\n\r\n--" . $boundary . "\r\n";
			$message.= "Content-type: text/plain;charset=utf-8\r\n\r\n";
			$message.= "
                A new post on [" . get_option('blogname') . "] by {$contributor->display_name} is pending review. \n
                URL : " . get_option('siteurl') . ". \n
                Author : {$contributor->user_login} <{$contributor->user_email}> (IP:{$_SERVER['REMOTE_ADDR']}). \n
                Title : $post_title. \n
                Powered by: WP Editorial Control
            ";
			$message.= "\r\n\r\n--" . $boundary . "\r\n";
			$message.= "Content-type: text/html;charset=utf-8\r\n\r\n";
			$message.= "
                <strong>A new post on [" . get_option('blogname') . "] by {$contributor->display_name} is pending review.</strong>
                <br/><br/>
                URL : " . get_option('siteurl') . "
                <br/>
                Author : {$contributor->user_login} <{$contributor->user_email}> (IP:{$_SERVER['REMOTE_ADDR']})
                <br/>
                Title : $post_title
            ";
			$category = get_the_category($post->ID);
			if (isset($category[0])) {
				$message.= " (Category : {$category[0]->name})<br/><br/>";
			}

			$message.= "
                Preview it: " . get_option('siteurl') . "/?p={$post->ID}&preview=true
                <br/>
                Edit/Publish it: " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID}
                <br/>
            ";
			$message.= "<br/>###";
			$post = get_post($post->ID);
			$message.= apply_filters('the_content', $post->post_content);
			$message.= "###<br/><br/>";
			$message.= "Edit/Publish it: " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID}<br/><br/><br/>";
			$message.= "Powered by: WP Editorial Control";
			$message.= "\r\n\r\n--" . $boundary . "--";
			wp_mail($emails, $subject, $message, $headers);
		}
	}
	elseif ($old_status == 'pending' && $new_status == 'publish' && $current_user->ID != $contributor->ID) {
		if (get_option('approvednotification') == 'yes') {
			$from_email = get_option('fromemail');
			$headers[] = 'From: Review Post <' . $from_email . '>';
			$headers[] = "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post->post_title . '" approved';
			$message = "{$contributor->display_name},\n\nCongratulations, your post has been approved and published at " . get_permalink($post->ID) . " .\n\n";
			$message.= "By {$current_user->display_name} <{$current_user->user_email}>\n\n\n";
			$message.= "Powered by: WP Editorial Control";
			wp_mail($contributor->user_email, $subject, $message, $headers);
		}
	}
	elseif ($old_status == 'pending' && $new_status == 'draft' && $current_user->ID != $contributor->ID) {
		if (get_option('declinednotification') == 'yes') {
			$from_email = get_option('fromemail');
			$headers[] = 'From: Review Post <' . $from_email . '>';
			$headers[] = "Content-Type: multipart/alternative;boundary=" . $boundary . "\r\n";
			$subject = '[' . get_option('blogname') . '] "' . $post->post_title . '" declined';
			$message = "{$contributor->display_name},\n\nSorry ,our post has not been approved. You can edit the post at " . get_option('siteurl') . "/wp-admin/post.php?action=edit&post={$post->ID} .\n\n";
			$message.= "By {$current_user->display_name} <{$current_user->user_email}>\n\n\n";
			$message.= "Powered by: Editorial Control";
			wp_mail($contributor->user_email, $subject, $message, $headers);
		}
	}
}